File: pythonlib/winfox/rdp/bridge.py
import asyncio
import json
import logging
import uuid
from typing import Any, Dict


logger = logging.getLogger(__name__)


class _ExtensionBridge:
    def __init__(self, port: int):
        self._port = port
        self._server = None
        self._ws = None
        self._pending: Dict[str, asyncio.Future] = {}
        self._connected = asyncio.Event()

    async def start(self):
        try:
            import websockets

            self._server = await websockets.serve(
                self._handler, "127.0.0.1", self._port
            )
            logger.info(f"Extension bridge listening on ws://127.0.0.1:{self._port}")
        except ImportError:
            logger.warning("websockets not installed, extension input unavailable")

    async def _handler(self, ws):
        self._ws = ws
        self._connected.set()
        logger.info("Extension connected")
        try:
            async for raw in ws:
                data = json.loads(raw)
                if data.get("type") == "hello":
                    logger.info(f"Extension hello: {data.get('extensionId')}")
                    continue
                msg_id = data.get("id")
                if msg_id and msg_id in self._pending:
                    self._pending[msg_id].set_result(data)
        except Exception:
            pass
        finally:
            for fut in list(self._pending.values()):
                if not fut.done():
                    fut.set_exception(ConnectionError("Extension bridge disconnected"))
            self._pending.clear()
            self._ws = None
            self._connected.clear()

    async def send_command(self, cmd: str, params: dict, timeout: float = 10.0) -> Any:
        if not self._ws:
            if not self._connected.is_set():
                try:
                    await asyncio.wait_for(self._connected.wait(), timeout=5.0)
                except asyncio.TimeoutError:
                    raise ConnectionError("Extension not connected")

        msg_id = str(uuid.uuid4())[:8]
        fut: asyncio.Future = asyncio.get_event_loop().create_future()
        self._pending[msg_id] = fut

        await self._ws.send(json.dumps({"id": msg_id, "cmd": cmd, "params": params}))

        try:
            result = await asyncio.wait_for(fut, timeout=timeout)
        finally:
            self._pending.pop(msg_id, None)

        if result.get("error"):
            raise RuntimeError(f"Extension error: {result['error']}")
        return result.get("result")

    async def stop(self):
        if self._ws:
            try:
                await self._ws.close()
            except Exception:
                pass
            self._ws = None
        for fut in list(self._pending.values()):
            if not fut.done():
                fut.set_exception(ConnectionError("Extension bridge stopped"))
        self._pending.clear()
        self._connected.clear()
        if self._server:
            self._server.close()
            await self._server.wait_closed()
            self._server = None

    @property
    def is_connected(self) -> bool:
        return self._ws is not None

File: packages/winfox/src/rdp/bridge.ts
import WebSocket, { WebSocketServer } from "ws";
import { v4 as uuidv4 } from "uuid";

export class ExtensionBridge {
  private port: number;
  private server: WebSocketServer | null = null;
  private ws: WebSocket | null = null;
  private pending: Map<string, { resolve: (val: any) => void; reject: (err: any) => void }> = new Map();
  private connectedResolvers: Array<() => void> = [];

  constructor(port: number) {
    this.port = port;
  }

  async start(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.server = new WebSocketServer({ host: "127.0.0.1", port: this.port }, () => {
          resolve();
        });
        this.server.on("connection", (ws) => this.handler(ws));
        this.server.on("error", (err) => reject(err));
      } catch (e) {
        reject(e);
      }
    });
  }

  private handler(ws: WebSocket) {
    this.ws = ws;
    while (this.connectedResolvers.length > 0) {
      const r = this.connectedResolvers.shift();
      if (r) r();
    }

    ws.on("message", (raw) => {
      try {
        const data = JSON.parse(raw.toString());
        if (data.type === "hello") return;
        const msgId = data.id;
        if (msgId && this.pending.has(msgId)) {
          this.pending.get(msgId)!.resolve(data);
        }
      } catch (e) {}
    });

    ws.on("close", () => {
      for (const [_, p] of this.pending.entries()) {
        p.reject(new Error("Extension bridge disconnected"));
      }
      this.pending.clear();
      if (this.ws === ws) {
        this.ws = null;
      }
    });
  }

  async sendCommand(cmd: string, params: any, timeoutSec: number = 10.0): Promise<any> {
    if (!this.ws) {
      await new Promise<void>((resolve, reject) => {
        const timer = setTimeout(() => reject(new Error("Extension not connected")), 5000);
        this.connectedResolvers.push(() => {
          clearTimeout(timer);
          resolve();
        });
      });
    }

    const msgId = uuidv4().substring(0, 8);
    let timeoutHandle: any;

    const prom = new Promise((resolve, reject) => {
      this.pending.set(msgId, { resolve, reject });
      timeoutHandle = setTimeout(() => {
        this.pending.delete(msgId);
        reject(new Error(`Extension command ${cmd} timed out after ${timeoutSec}s`));
      }, timeoutSec * 1000);
    });

    this.ws!.send(JSON.stringify({ id: msgId, cmd, params }));
    try {
      const result: any = await prom;
      if (result.error) throw new Error(`Extension error: ${result.error}`);
      return result.result;
    } finally {
      clearTimeout(timeoutHandle);
      this.pending.delete(msgId);
    }
  }

  async stop(): Promise<void> {
    if (this.ws) {
      try { this.ws.close(); } catch (e) {}
      this.ws = null;
    }
    for (const [_, p] of this.pending.entries()) {
      p.reject(new Error("Extension bridge stopped"));
    }
    this.pending.clear();
    if (this.server) {
      this.server.close();
      this.server = null;
    }
  }

  get isConnected(): boolean {
    return this.ws !== null && this.ws.readyState === WebSocket.OPEN;
  }
}
